From 55c49094b163fb5226f3e6af26b858e030e3f70b Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Thu, 5 Jun 2025 23:25:59 +0800
Subject: [PATCH 066/105] mtk: mt76: mt7996: add 5th rx rssi dump support

Dump the rssi of 5th rx for mt7990 6G 4T5R & mt7992 5G 5T5R

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>

/home/mtk21047/ws/openwrt/20250610/openwrt/openwrt/build_dir/target-aarch64_cortex-a53_mus
l/linux-mediatek_filogic/mt76-2025.06.01~79dd14f2/mt76.h:865:1: note: offset of packed bit
-field 'band' has changed in GCC 4.4
  865 | } __packed;
      | ^

Change to u16 to get rid of this compiler note.

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>

When there is no STA connected, the RSSI signal of a frame calculated
by mt76_rx_signal becomes 0, as the RSSI of the 5th RX is set to 0.
Therefore, set the RSSI of the 5th RX to the mininum value = -128 to avoid
polluting the RSSI signal.
For example, the auth and assoc resp frame sent from the STA will have
an RSSI of 0, which causes the rssi_reject_assoc_rssi
check to fail due to inaccurate rssi values.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt76.h           |  3 +++
 mt7996/debugfs.c | 47 +++++++++++++++++++++++++++++++++++++----------
 mt7996/mac.c     | 13 ++++++++++---
 3 files changed, 50 insertions(+), 13 deletions(-)

diff --git a/mt76.h b/mt76.h
index 9b36a0c7b..e2fe1b038 100644
--- a/mt76.h
+++ b/mt76.h
@@ -826,9 +826,11 @@ struct mt76_rx_status {
 	u32 ampdu_ref;
 	u32 timestamp;
 	u8 iv[6];
+
 	u16 freq:13, phy_idx:2, aggr:1;
 	u8 qos_ctl;
 	u16 seqno;
+
 	u32 flag;
 	u8 enc_flags;
 	u8 encoding:3, bw:4;
@@ -843,6 +845,7 @@ struct mt76_rx_status {
 			u8 gi:2;
 		} eht;
 	};
+
 	u16 amsdu:1, first_amsdu:1, last_amsdu:1;
 	u16 nss:3, band:3;
 	u16 rate_idx:7;
diff --git a/mt7996/debugfs.c b/mt7996/debugfs.c
index 6551995f1..c2a2bbfd8 100644
--- a/mt7996/debugfs.c
+++ b/mt7996/debugfs.c
@@ -1862,6 +1862,8 @@ mt7996_link_sta_info_show(struct seq_file *file, void *data)
 	struct mt7996_sta_link *msta_link;
 	struct mt76_sta_stats *stats;
 	struct mt76_wcid *wcid;
+	struct mt7996_phy *phy;
+	unsigned long rx_mask;
 	char buf[100];
 	u8 ac;
 
@@ -1874,21 +1876,46 @@ mt7996_link_sta_info_show(struct seq_file *file, void *data)
 	}
 	wcid = &msta_link->wcid;
 	stats = &wcid->stats;
+	phy = __mt7996_phy(dev, wcid->phy_idx);
+	rx_mask = mt7996_rx_chainmask(phy);
 
 	seq_printf(file, "WCID: %hu\n", wcid->idx);
 	seq_printf(file, "Link ID: %hhu\n", link_sta->link_id);
 	seq_printf(file, "Link Address: %pM\n", link_sta->addr);
 	seq_printf(file, "Status:\n");
-	seq_printf(file, "\tRSSI: %d [%hhd, %hhd, %hhd, %hhd] dBm\n",
-		   msta_link->signal, msta_link->chain_signal[0], msta_link->chain_signal[1],
-		   msta_link->chain_signal[2], msta_link->chain_signal[3]);
-	seq_printf(file, "\tACK RSSI: %d [%hhd, %hhd, %hhd, %hhd] dBm\n",
-		   msta_link->ack_signal, msta_link->chain_ack_signal[0],
-		   msta_link->chain_ack_signal[1], msta_link->chain_ack_signal[2],
-		   msta_link->chain_ack_signal[3]);
-	seq_printf(file, "\tACK SNR: [%hhd, %hhd, %hhd, %hhd] dBm\n",
-		   msta_link->chain_ack_snr[0], msta_link->chain_ack_snr[1],
-		   msta_link->chain_ack_snr[2], msta_link->chain_ack_snr[3]);
+
+#define dump_chain_signal(type, buf, buflen, sta_link, rx_mask)			\
+	do {									\
+		char *pos = (buf);						\
+		int chain;							\
+		size_t left = (buflen);						\
+		int n;								\
+		n = snprintf(pos, left, "[");					\
+		pos += n; left -= n;						\
+		for_each_set_bit(chain, &(rx_mask), IEEE80211_MAX_CHAINS) {	\
+			if (chain) {						\
+				n = snprintf(pos, left, ", ");			\
+				pos += n; left -= n;				\
+			}							\
+			n = snprintf(pos, left, "%hhd",				\
+				(sta_link)->chain_##type[(chain)]);		\
+			pos += n; left -= n;					\
+		}								\
+		snprintf(pos, left, "] dBm\n");					\
+	} while (0)
+
+	dump_chain_signal(signal, buf, sizeof(buf), msta_link, rx_mask);
+	seq_printf(file, "\tRSSI: %d %s", msta_link->signal, buf);
+
+	dump_chain_signal(ack_signal, buf, sizeof(buf), msta_link, rx_mask);
+	seq_printf(file, "\tACK RSSI: %d %s", msta_link->ack_signal, buf);
+
+	/* SNR of 5th rx is not supported */
+	rx_mask &= ~BIT(4);
+	dump_chain_signal(ack_snr, buf, sizeof(buf), msta_link, rx_mask);
+	seq_printf(file, "\tACK SNR: %s", buf);
+#undef dump_chain_signal
+
 	seq_printf(file, "Rate:\n");
 
 	mt7996_parse_rate(&wcid->rate, buf, sizeof(buf));
diff --git a/mt7996/mac.c b/mt7996/mac.c
index 36db0473d..3fc2eb104 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -484,21 +484,28 @@ mt7996_mac_fill_rx(struct mt7996_dev *dev, enum mt76_rxq_id q,
 
 		v3 = le32_to_cpu(rxv[3]);
 
-		status->chains = mphy->antenna_mask;
+		status->chains = mt7996_rx_chainmask(phy);
 		status->chain_signal[0] = to_rssi(MT_PRXV_RCPI0, v3);
 		status->chain_signal[1] = to_rssi(MT_PRXV_RCPI1, v3);
 		status->chain_signal[2] = to_rssi(MT_PRXV_RCPI2, v3);
 		status->chain_signal[3] = to_rssi(MT_PRXV_RCPI3, v3);
+		/* no 5th rx info in rxd for data frame */
+		status->chain_signal[4] = -128;
 
 		if (msta_link) {
+			u16 rx_mask = status->chains;
 			int i;
 
+			/* use the value read from rxv for 5th rx */
+			if (rx_mask & BIT(4))
+				status->chain_signal[4] = msta_link->chain_signal_wf4;
+
 			memcpy(msta_link->chain_signal, status->chain_signal,
 			       IEEE80211_MAX_CHAINS);
-			msta_link->signal = mt76_rx_signal(mphy->antenna_mask,
+			msta_link->signal = mt76_rx_signal(rx_mask,
 							   msta_link->chain_signal);
 
-			for (i = 0; i < IEEE80211_MAX_CHAINS; ++i)
+			for (i = 0; i < hweight16(rx_mask); ++i)
 				ewma_avg_signal_add(msta_link->chain_signal_avg + i,
 						    -msta_link->chain_signal[i]);
 			ewma_avg_signal_add(&msta_link->signal_avg, -msta_link->signal);
-- 
2.45.2

